### DISCUSSION TITLE
Would a putative a conio lib for the Atari 7800 in 160B mode only provide 64 tiles for all colors?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1733] [Q&A]
@karrika I wonder why you decided to use the 160A mode instead of the more colorful 160B mode for the conio driver in 160x240 resolution.
I guess it's  because 4 bits per pixel would imply only 64 (256/4) tiles for all colors, which would paradoxically imply only one color.
Am I right?

In the future I may try to implement something like a 160B conio lib (at least for my own use) because the user could provide and link a different font/tile file in 12+1 colors.

How much different would the code be compared to yours? I supposed some initial setup would change and the display lists would be different. Of course the encoding of the tiles is totally different and more convoluted for the 160B mode.

**Comment by karrika** [comment]
Yep. The problem here is really that we cannot use more than 8 lines per character with inirect addressing.
It may be better to take the tgi route for colours and skip conio.

**Comment by Fabrizio-Caruso** [comment]
I have no idea how full TGI graphics could be possible with just 4k of RAM. Maybe you mean TGI (sort of) sprites as for the Atari Lynx but with no other TGI apis that assume a bitmap mode, which requires a huge framebuffer. I guess the tile/sprite data would still need to be in ROM. I am still not familiar enough with the way the A7800 is programmed despite some efforts. I need to keep reading about it.
Nevertheless for my very specific usecase even a conio-like approach with just 64 8x8 "shapes" could be of some use. The ideal solution would be to have 256 8x8 shapes but there may just be no easy way to do it.
In Cross-Lib hardware-abstracted graphics I assume the screen is made of cells. So it is inherently indirect. Each cell of the screen has a value that is an index that points to its shape.

**Reply by karrika** [reply]
The zones needed for a 20x25 pixel zone list eats up all RAM1. But you would have 21 colours for the pixels. The program ram would then be RAM3 and the C stack RAM2.

**Comment by Fabrizio-Caruso** [comment]
I am not implementing any specific game where I know in advance how many sprites I will need. I implement a tile-based with at least 64 "shapes" possibly with colors (which is the bare common minimum on all 200 systems). If a specific system has a bitmap mode and no tile/character mode, I use the bitmap mode to implement tiles. If it doesn't I use its own tile/text mode.
The big problem I see with the Atari 7800 is the 4k ram limit, which cannot simply hold a framebuffer for each pixel in 4-bit mode (12+1 colors). I know some cartridge provide a ram expansion. If the ram expansion can be used for graphics (e.g., as a framebuffer), then there is a solution (sort of).

**Reply by karrika** [reply]
You can have carts with extra RAM.

**Reply by karrika** [reply]
The TGI driver would not make your life easier as it does now handle sprites.

My feeling is that you just have to maintain your own display lists and zones. A zone can also be pictured as a tile. Instead of having an index on a screen buffer that points to a sprite you can make the zone point to the graphics directly. The tricky part begins when the graphics moves from one line to the next as it moves from one zone to the next.

On the other hand the conio implementation has exactly what you wanted. 128 sprites in colour. You can always just move them in memory by accessing screen[index] directly. Much faster than pretending they are text.
